src/librustc_mir/build/matches/mod.rs

```diff
@@ -99,6 +124,61 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                 .zip(pre_binding_blocks.iter().zip(pre_binding_blocks.iter().skip(1)))
                 .map(|((arm_index, pattern, guard),
                        (pre_binding_block, next_candidate_pre_binding_block))| {
+
+                    if let (true, Some(borrow_temp)) = (tcx.emit_read_for_match(),
+                                                        borrowed_input_temp.clone()) {
+                        // inject a fake read of the borrowed input at
+                        // the start of each arm's pattern testing
+                        // code.
+                        //
+                        // This should ensure that you cannot change
+                        // the variant for an enum while you are in
+                        // the midst of matching on it.
+                        //
+                        // FIXME: I originally had put this at the
+                        // start of each elem of arm_blocks (see
```

Comment:
The theoretically "right" place to put this is "after all arms", just before the `unreachable` of the `FalseEdges`, where we witness the exhaustiveness of the match.

Note that this (as well as your code, and soundly) allows for mutating the discriminant in a guard that later *diverges*:
```Rust
let mut foo = Some(4);
match foo {
    _ if { if maybe {
        // modify the guard, then *diverge*
        foo = None;
        panic!()
    } else { false } } => {},
    _ => {}
}
```

1. The `unreachable` at the end of the false-edges chain. That's the "canonical" `unreachable` for borrowck purposes, but is not accessible to LLVM:
    https://github.com/pnkfelix/rust/blob/ea9afcb6a7c49752271652d40ca7d365057c82fc/src/librustc_mir/build/matches/mod.rs#L195
2. The `unreachable` at the end of the real edges chain. That's the "canonical" `unreachable` for trans purposes:
    https://github.com/pnkfelix/rust/blob/ea9afcb6a7c49752271652d40ca7d365057c82fc/src/librustc_mir/build/matches/mod.rs#L215
3. An `unreachable` at the end of the true edge of each guard that is in unreachable code
    https://github.com/pnkfelix/rust/blob/ea9afcb6a7c49752271652d40ca7d365057c82fc/src/librustc_mir/build/matches/mod.rs#L585

The `unureachable` at (3) can only be reached through a false edge, and I think we can replace it by going to the same place that the false edge goes - i.e. to `candidate.next_candidate_pre_binding_block`, and therefore eventualy to a (2) `unreachable` - to avoid adding "unnecessary" unreachable terminators.

At every point in "user code" the (2) or (3) `unreachable` can be reached, following the false edges will reach the (1) `unreachable`, so we can add the test there.

Reply:
@arielb1 I'm having a little trouble parsing your comment at the top here.

I *think* that when you wrote:

> Note that this (as well as your code, and soundly) allows for mutating the discriminant in a guard that later diverges

the "this" in that sentence is referencing your original (now struckout) suggestion of

> ~~The theoretically "right" place to put this is "after all arms", just before the unreachable of the FalseEdges~~

(and then the rest of that text, including the enumerated list with three different `unreachable's`, was you working out a justification for the struckout suggestion, right?) 

Is the above understanding correct?

----

And so the summary conclusion is ... we certainly cannot rely on just injecting the `ReadForMatch` one time at the end of the whole `match` (at one of the unreachable points), because we need to retain integrity of the data structure in the face of arms that diverge.

The examples in your most recent comments: Those are meant to be concrete tests showing that the currently implemented strategy (of injecting a `ReadForMatch` at the start of each arm) is **necessary*.*, right? I've been trying to figure out if either of them are illustrating a flaw/oversight in the current strategy.